src/Nix/Type/Infer.hs

```diff
@@ -596,14 +596,14 @@ normalizeScheme (Forall _ body) = Forall (fmap snd ord) (normtype body)
 ---------------------------------------------------------------------------------
 
 newtype Solver m a = Solver (LogicT (StateT [TypeError] m) a)
-    deriving (Functor, Applicative, Alternative, Monad, MonadPlus,
+    deriving (Functor, Applicative, Alternative, Monad,
```

Comment:
Lets preserve `MonadPlus` here.

This conclusion derived from the below
```haskell
throwError err = Solver $ lift (modify (err :)) *> Applicative.empty
```
thinking through the code.
`MonadPlus` does computational saving between the `unifyMany` & `unifies` butual recursion:
```haskell
unifyMany :: Monad m => [Type] -> [Type] -> Solver m Subst
...
unifyMany (t1 : ts1) (t2 : ts2) = do
  su1 <- unifies t1 t2
  su2 <- unifyMany (apply su1 ts1) (apply su1 ts2)
  pure (su2 `compose` su1)

...
unifies (TList xs) (TList ys)
  | allSameType xs && allSameType ys = case (xs, ys) of
    (x : _, y : _) -> unifies x y
    _              -> pure emptySubst
  | length xs == length ys = unifyMany xs ys
unifies (t1 :~> t2) (t3 :~> t4) = unifyMany [t1, t2] [t3, t4]
```